core/jobs/batch_jobs/store_profile_images_to_gcs_jobs.py

```diff
@@ -0,0 +1,332 @@
+                self._make_profile_picture_valid)
+        )
+
+        write_png_files_to_gcs = (
+            users_with_valid_username
+            | 'Map files for png' >> beam.Map(self._generate_png_file_object)
+            | 'Write png file to GCS' >> gcs_io.WriteFile(mime_type='image/png')
+        )
+
+        total_png_images = (
+            write_png_files_to_gcs
+            | 'Total png images wrote to GCS' >> (
+                job_result_transforms.CountObjectsToJobRunResult(
+                    'TOTAL PNG IMAGES'))
+        )
+
+        write_webp_files_to_gcs = (
+            users_with_valid_username
+            | 'Map files for webp' >> beam.Map(self._generate_webp_file_object)
+            | 'Write webp file to GCS' >> gcs_io.WriteFile(
+                mime_type='image/webp')
+        )
+
+        total_webp_images = (
+            write_webp_files_to_gcs
+            | 'Total webp images wrote to GCS' >> (
+                job_result_transforms.CountObjectsToJobRunResult(
+                    'TOTAL WEBP IMAGES'))
+        )
+
+        unused_put_results = (
+            users_with_valid_username
+            | 'Updating the datastore with valid profile images'
+            >> ndb_io.PutModels()
+        )
+
+        return (
+            (
+                total_png_images,
+                total_webp_images
+            )
+            | 'Combine results' >> beam.Flatten()
+        )
+
+
+class AuditProfilePictureFromGCSJob(base_jobs.JobBase):
+    """Audit profile pictures are present in GCS."""
+
+    def _png_base64_to_webp_base64(self, png_base64: str) -> str:
+        """Convert png base64 to webp base64.
+
+        Args:
+            png_base64: str. The png base64 string.
+
+        Returns:
+            str. The webp base64 string.
+        """
+        png_binary = utils.convert_png_data_url_to_binary(png_base64)
+        output = io.BytesIO()
+        image = Image.open(io.BytesIO(png_binary)).convert('RGB')
+        image.save(output, 'webp')
+        webp_binary = output.getvalue()
+        return utils.convert_png_or_webp_binary_to_data_url(
+            webp_binary, 'webp')
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        users_with_valid_username = (
+            self.pipeline
+            | 'Get all non-deleted UserSettingsModel' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all(include_deleted=False))
+            | 'Filter valid users with not None username' >> beam.Filter(
+                lambda model: model.username is not None)
+        )
+
+        username_with_profile_data = (
+            users_with_valid_username
+            | 'Map username and data url' >> beam.Map(
+                lambda model: (
+                    model.username,
+                    model.profile_picture_data_url.replace(
+                        '%2B', '+').replace('%3D', '=').replace('%0A', '')))
```

Comment:
The job errored on the server because newly created user settings models may have a valid username but an empty profile_picture_data_url field -- the image may be stored on GCS directly. The highlighted statement throws an error because "replace" method is accessed on a None type object (profile_picture_data_url) in the case of newly created models.

We will need to update this audit job to correctly handle that scenario.

Reply:
Raised the PR - https://github.com/oppia/oppia/pull/18142
Thanks